src/member.rs

```diff
@@ -332,77 +326,80 @@ where
             debug!(target: "rush-member", "Unit of non-zero round with only {:?} parents while at least {:?} are required.", n_parents, threshold);
             return false;
         }
-        let control_hash = &su.unit.inner.control_hash;
-        if round > 0 && !control_hash.parents[su.creator()] {
+        let control_hash = &pre_unit.control_hash;
+        if round > 0 && !control_hash.parents[pre_unit.creator()] {
             debug!(target: "rush-member", "Unit does not have its creator's previous unit as parent.");
             return false;
         }
         true
     }
 
-    fn validate_unit(&self, su: &SignedUnit<H, D, Signature>) -> bool {
+    fn validate_unit(&self, su: &SignedUnit<'a, H, D, KB>) -> bool {
         // TODO: make sure we check all that is necessary for unit correctness
         // TODO: consider moving validation logic for units and alerts to another file, note however
         // that access to the authority list is required for validation.
-        if su.unit.session_id != self.config.session_id {
+        if su.signed().session_id != self.config.session_id {
             // NOTE: this implies malicious behavior as the unit's session_id
             // is incompatible with session_id of the message it arrived in.
-            debug!(target: "rush-member", "A unit with incorrect session_id! {:?}", su);
+            debug!(target: "rush-member", "A unit with incorrect session_id! {:?}", su.unchecked);
             return false;
         }
-        if !su.verify_signature(&self.keybox) {
-            debug!(target: "rush-member", "A unit with incorrect signature! {:?}", su);
+        if !su.verify() {
```

Comment:
Shouldn't the unit already by verified by the virtue of its type?

Reply:
Fair point